libhb: use the hwaccel decoders in scan, and enable VideoToolbox in the UI.

galad87 opened the pull request:
…n actually decode the video or not. Hwaccel will fall back automatically to software, and we need to know if it will before starting a job.

win ui and linux ui will need to hook up the option for nvdec in hb_scan if they want to allow it.

QSV uses a separate decoder, and not hwaccel, so it might not require this and has not been modified yet.


**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [ ] Ubuntu Linux

galad87 committed:
libhb: use the hwaccel decoders in scan, to check if they hardware can actually decode the video or not. Hwaccel will fall back automatically to software, and we need to know if it will before starting a job.

sr55 commented:
Worth adding a log message to indicate what is detected? 

On or off, I see no difference in behaviour for the test files I ran.  Not seeing any GPU activity but it may be too short to show. (With it set to 1)

galad87 commented:
It prints "hwaccel" if NVDEC works:

>  scan: supported video decoders: avcodec hwaccel

galad87 committed:
libhb: add a special filter wrapper at the being of the filters chain to handle rotation and size changes.

galad87 commented:
Added a new VideoToolbox "prefilter" filter to handle size and rotation changes. The VideoToolbox crop & scale and rotate filters are not AVFilters, so they can't be easily used in decavcodec.c like the others, and modifying the filters chain dynamically will require some additional work, so for now I added a small optional filter chain inside the main filters chain 😵‍💫. IMHO we should remove all the filters from decavcodec.c in the future.

Anyway, this was the last improvement required, and so there is a new checkbox to enable the hardware decoders in the Advanced preferences, only for the HW to HW path for now, but the HW to CPU works well too, it just requires a few additional UI changes.

galad87 committed:
MacGui: add checkboxes to enable the videotoolbox hardware decoders always or only in full path.

galad87 committed:
libhb: disable job->hw_decode check in work.c until the winui is updated to use nvdec during scan.
